webalizer.1
.TH webalizer 1 "22-Oct-2001" "Version 2.01" "The Webalizer"
.SH NAME
webalizer - A web server log file analysis tool.
.SH SYNOPSIS
.B webalizer
[\fI option ... \fP] [\fI log-file \fP]
.PP
.B webazolver
[\fI option ... \fP] [\fI log-file \fP]
.PP
.SH DESCRIPTION
\fB-----------------------------------------------------------------------
This page is not being actively maintained for Stone Steps Webalizer. 
See the README file for for up-to-date documentation.
-----------------------------------------------------------------------\fP
.PP
The \fIWebalizer\fP is a web server log file analysis program which produces
usage statistics in HTML format for viewing with a browser.  The results
are presented in both columnar and graphical format, which facilitates
interpretation.  Yearly, monthly, daily and hourly usage statistics are
presented, along with the ability to display usage by site, URL, referrer,
user agent (browser), username, search strings, entry/exit pages,  and
country (some information may not be available if not present in the log
file being processed).
.PP
The \fIWebalizer\fP supports \fBCLF\fP (common log format) log files,
as well as \fBCombined\fP log formats as defined by NCSA and others,
and variations of these which it attempts to handle intelligently.  In
addition, the \fIWebalizer\fP also supports \fIwu-ftpd\fP \fBxferlog\fP
formatted log files, allowing analysis of ftp servers, and
\fBsquid\fP proxy logs.  Logs may also be compressed, via \fBgzip\fP.
If a compressed log file is detected, it will be automatically uncompressed
while it is read.  Compressed logs must have the standard \fIgzip\fP
extension of \fB.gz\fP.
.PP
\fIwebazolver\fP is normally just a symbolic link to the \fIwebalizer\fP.
When run as \fIwebazolver\fP, only DNS file creation/updates are performed,
and the program will exit once complete.  All normal options and
configuration directives are available, however many will not be used.
In addition, a DNS cache file must be specified.  If the number of DNS
children processes to use are not specified, the \fIwebazolver\fP will
default to \fB5\fP.
.PP
This documentation applies to The Webalizer Version 2.01
.SH RUNNING THE WEBALIZER
The \fIWebalizer\fP was designed to be run from a Unix command line prompt or
as a \fBcrond(8)\fP job. Once executed, the general flow of the program is:
.TP 8
.B o
A default configuration file is scanned for.  A file named
\fIwebalizer.conf\fP is searched for in the current directory, and if
found, it's configuration data is parsed.  If the file is not
present in the current directory,  the file \fI/etc/webalizer.conf\fP
is searched for and, if found, is used instead.
.TP 8
.B o
Any command line arguments given to the program are parsed.  This
may include the specification of a configuration file, which is
processed at the time it is encountered.
.TP 8
.B o
If a log file was specified, it is opened and made ready for
processing.  If no log file was given, \fISTDIN\fP is used for input.
If the log filename '\fB-\fP' is specified, \fISTDIN\fP will be forced.
.TP 8
.B o
If an output directory was specified, the program does a \fBchdir(2)\fP to
that directory in prepration for generating output.  If no output
directory was given, the current directory is used.
.TP 8
.B o
If a non-zero number of DNS Children processes were specified, they will
be started, and the specified log file will be processed, creating or
updating the specified DNS cache file.
.TP 8
.B o
If no hostname was given, the program attempts to get the hostname
using a \fBuname(2)\fP system call.  If that fails, \fIlocalhost\fP
is used.
.TP 8
.B o
A history file is searched for in the current directory (output
directory) and read if found.  This file keeps totals for previous
months, which is used in the main \fIindex.html\fP HTML document.
.B Note:
The file location can now be specified with the \fIHistoryName\fP
configuration option.
.TP 8
.B o
If incremental processing was specified, a data file is searched for
and loaded if found, containing the 'internal state' data of the
program at the end of a previous run.
.B Note:
The file location can now be specified with the \fIIncrementalName\fP
configuration option.
.TP 8
.B o
Main processing begins on the log file.  If the log spans multiple
months, a seperate HTML document is created for each month.
.TP 8
.B o
After main processing, the main \fIindex.html\fP page is created, which
has totals by month and links to each months HTML document.
.TP 8
.B o
A new history file is saved to disk, which includes totals generated
by The Webalizer during the current run.
.TP 8
.B o
If incremental processing was specified, a data file is written that
contains the 'internal state' data at the end of this run.
.SH INCREMENTAL PROCESSING
Version 1.2x of The \fIWebalizer\fP adds incremental run capability.  Simply
put, this allows processing large log files by breaking them up into
smaller pieces, and processing these pieces instead.  What this means
in real terms is that you can now rotate your log files as often as you
want, and still be able to produce monthly usage statistics without the
loss of any detail.  Basically, The \fIWebalizer\fP saves and restores all
internal data in a file named \fIwebalizer.current\fP.  This allows the
program to 'start where it left off' so to speak, and allows the
preservation of detail from one run to the next.  The data file is
placed in the current output directory, and is a plain ascii text
file that can be viewed with any standard text editor.  It's location
and name may be changed using the \fIIncrementalName\fP configuration
keyword.
.PP
Some special precautions need to be taken when using the incremental
run capability of The \fIWebalizer\fP.  Configuration options should not be
changed between runs, as that could cause corruption of the internal
data stored.  For example, changing the \fIMangleAgents\fP level will cause
different representations of user agents to be stored, producing invalid
results in the user agents section of the report.  If you need to change
configuration options, do it at the end of the month after normal
processing of the previous month and before processing the current month.
You may also want to delete the \fIwebalizer.current\fP file as well.
.PP
The \fIWebalizer\fP also attempts to prevent data duplication by keeping
track of the timestamp of the last record processed.  This timestamp
is then compared to current records being processed, and any records
that were logged previous to that timestamp are ignored.  This, in
theory, should allow you to re-process logs that have already been
processed, or process logs that contain a mix of processed/not yet
processed records, and not produce duplication of statistics.  The
only time this may break is if you have duplicate timestamps in two
seperate log files... any records in the second log file that do have
the same timestamp as the last record in the previous log file processed,
will be discarded as if they had already been processed.  There are
lots of ways to prevent this however, for example, stopping the web
server before rotating logs will prevent this situation.  This setup
also necessitates that you always process logs in chronological order,
otherwise data loss will occur as a result of the timestamp compare.
.SH REVERSE DNS LOOKUPS
The Webalizer supports reverse DNS lookups through a \fIDNS cache file\fP
that is either created/updated at run-time, or has been previously
created, either by a previous run of the \fIwebalizer\fP, or by running
the stand-alone version, \fIwebazolver\fP.  In order to perform reverse
DNS lookups, a \fBDNSCache\fP filename must be specified.  In order to
create/update the cache file at run-time, the \fBDNSChildren\fP number
must be non-zero.  The \fBDNSChildren\fP value specifies the number of
children processes to fork, each of which will perform reverse DNS
lookups in order to create/update the DNS cache file.  See the file
\fIDNS.README\fP for additional information.
.SH COMMAND LINE OPTIONS
The Webalizer supports many different configuration options that will
alter the way the program behaves and generates output.  Most of these
can be specified on the command line, while some can only be specified
in a configuration file. The command line options are listed below,
with references to the corresponding configuration file keywords.
.PP
.I General Options
.TP 8
.B \-h
Display all available command line options and exit program.
.TP 8
.B \-v \-V
Display program version and exit program.
.TP 8
.B \-d
\fBDebug\fP.  Display debugging information for errors and warnings.
.TP 8
.B \-i
\fBIgnoreHist\fP.  Ignore history.  \fBUSE WITH CAUTION\fP. This
will cause The \fIWebalizer\fP to ignore any previous monthly history
file only.  Incremental data (if present) is still processed.
.TP 8
.B \-p
\fBIncremental\fP.  Preserve internal data between runs.
.TP 8
.B \-q
\fBQuiet\fP.  Supress informational messages.  Does not supress
warnings or errors.
.TP 8
.B \-Q
\fBReallyQuiet\fP.  Supress all messages including warnings and errors.
.TP 8
.B \-T
\fBTimeMe\fP.  Force display of timing information at end of processing.
.TP 8
.B \-c \fIfile\fP
Use configuration file \fIfile\fP.
.TP 8
.B \-n \fIname\fP
\fBHostName\fP.  Use the hostname \fIname\fP.
.TP 8
.B \-o \fIdir\fP
\fBOutputDir\fP.  Use output directory \fIdir\fP.
.TP 8
.B \-t \fIname\fP
\fBReportTitle\fP.  Use \fIname\fP for report title.
.TP 8
.B \-F \fP( \fBc\fPlf | \fBf\fPtp | \fBs\fPquid )
\fBLogType\fP.  Specify log type to be processed.  Value can be either
\fIc\fPlf, \fIf\fPtp or \fIs\fPquid format.  If not specified, will
default to \fBCLF\fP format.  \fIFTP\fP logs must be in standard
wu-ftpd \fIxferlog\fP format.
.TP 8
.B \-f
\fBFoldSeqErr\fP.  Fold out of sequence log records back into analysis,
by treating as if they were the same date/time as the last good record.
Normally, out of sequence log records are simply ignored.
.TP 8
.B \-Y
\fBCountryGraph\fP. Supress country graph.
.TP 8
.B \-G
\fBHourlyGraph\fP.  Supress hourly graph.
.TP 8
.B \-x \fIname\fP
\fBHTMLExtension\fP.  Defines HTML file extension to use.  If not
specified, defaults to \fIhtml\fP.  Do not include the leading
period.
.TP 8
.B \-H
\fBHourlyStats\fP.  Supress hourly statistics.
.TP 8
.B \-L
\fBGraphLegend\fP.  Supress color coded graph legends.
.TP 8
.B \-l \fInum\fP
\fBGraphLines\fP.  Specify number of background lines. Default
is 2.  Use zero ('0') to disable the lines.
.TP 8
.B \-P \fIname\fP
\fBPageType\fP.  Specify file extensions that are considered \fIpages\fP.
Sometimes referred to as \fIpageviews\fP.
.TP 8
.B \-m \fInum\fP
\fBVisitTimeout\fP.  Specify the Visit timeout period.  Specified in
number of seconds.  Default is 1800 seconds (30 minutes).
.TP 8
.B \-I \fIname\fP
\fBIndexAlias\fP.  Use the filename \fIname\fP as an additional alias
for \fIindex.\fP.
.TP 8
.B \-M \fInum\fP
\fBMangleAgents\fP.  Mangle user agent names according to the mangle
level specified by \fInum\fP.  Mangle levels are:
.TP 12
.ti 8
.B 5
Browser name and major version.
.TP 12
.ti 8
.B 4
Browser name, major and minor version.
.TP 12
.ti 8
.B 3
Browser name, major version, minor version to two decimal places.
.TP 12
.ti 8
.B 2
Browser name, major and minor versions and sub-version.
.TP 12
.ti 8
.B 1
Browser name, version and machine type if possible.
.TP 12
.ti 8
.B 0
All informaiton (left unchanged).
.TP 8
.B \-g \fInum\fP
\fBGroupDomains\fP. Automatically group sites by domain.  The
grouping level specified by \fInum\fP can be thought of as 'the
number of dots' to display in the grouping.  The default value
of \fB0\fP disables any domain grouping.
.TP 8
.B \-D \fIname\fP
\fBDNSCache\fP.  Use the DNS cache file \fIname\fP.
.TP 8
.B \-N \fInum\fP
\fBDNSChildren\fP.  Use \fInum\fP DNS children processes to perform DNS
lookups, either creating or updateing the DNS cache file.  Specify zero
(\fB0\fP) to disable cache file creation/updates.  If given, a DNS cache
filename must be specified.
.PP
.I Hide Options
.TP 8
.B \-a \fIname\fP
\fBHideAgent\fP.  Hide user agents matching \fIname\fP.
.TP 8
.B \-r \fIname\fP
\fBHideReferrer\fP.  Hide referrer matching \fIname\fP.
.TP 8
.B \-s \fIname\fP
\fBHideSite\fP.  Hide site matching \fIname\fP.
.TP 8
.B \-X \fIname\fP
\fBHideAllSites\fP.  Hide all individual sites (only display groups).
.TP 8
.B \-u \fIname\fP
\fBHideURL\fP.  Hide URL matching \fIname\fP.
.PP
.I Table size options
.TP 8
.B \-A \fInum\fP
\fBTopAgents\fP.  Display the top \fInum\fP user agents table.
.TP 8
.B \-R \fInum\fP
\fBTopReferrers\fP.  Display the top \fInum\fP referrers table.
.TP 8
.B \-S \fInum\fP
\fBTopSites\fP.  Display the top \fInum\fP sites table.
.TP 8
.B \-U \fInum\fP
\fBTopURLs\fP.  Display the top \fInum\fP URL's table.
.TP 8
.B \-C \fInum\fP
\fBTopCountries\fP.  Display the top \fInum\fP countries table.
.TP 8
.B \-e \fInum\fP
\fBTopEntry\fP.  Display the top \fInum\fP entry pages table.
.TP 8
.B \-E \fInum\fP
\fBTopExit\fP.  Display the top \fInum\fP exit pages table.
.SH CONFIGURATION FILES
Configuration files are standard \fBascii(7)\fP text files that may be created
or edited using any standard editor.  Blank lines and lines that begin
with a pound sign ('#') are ignored.  Any other lines are considered to
be configurgation lines, and have the form "Keyword Value", where the
\'Keyword\' is one of the currently available configuration keywords defined
below, and 'Value' is the value to assign to that particular option.  Any
text found after the keyword up to the end of the line is considered the
keyword's value, so you should not include anything after the actual value
on the line that is not actually part of the value being assigned.  The
file \fIsample.conf\fP provided with the distribution contains lots of useful
documentation and examples as well.

.I General Configuration Keywords
.TP 8
.B LogFile \fIname\fP
Use log file named \fIname\fP.  If none specified, \fISTDIN\fP will be used.
.TP 8
.B LogType \fIname\fP
Specify log file type as \fIname\fP. Values can be either \fIweb\fP,
\fIs\fPquid or \fIftp\fP, with the default being \fBweb\fP.
.TP 8
.B OutputDir \fIdir\fP
Create output in the directory \fIdir\fP.  If none specified, the current
directory will be used.
.TP 8
.B HistoryName \fIname\fP
Filename to use for history file.  Relative to output directory unless
absolute name is given (ie: starts with '/'). Defaults to
\'\fBwebalizer.hist\fP' in the standard output directory.
.TP 8
.B ReportTitle \fIname\fP
Use the title string \fIname\fP for the report title.  If none
specified, use the default of (in english) "\fIUsage Statistics for \fP".
.TP 8
.B Hostname \fIname\fP
Set the hostname for the report as \fIname\fP.  If none specified, an
attempt will be made to gather the hostname via a \fBuname(2)\fP system
call.  If that fails, \fIlocalhost\fP will be used.
.TP 8
.B UseHTTPS \fP( yes | \fBno\fP )
Use \fIhttps://\fP on links to URLS, instead of the default \fIhttp://\fP,
in the '\fBTop URL's\fP' table.
.TP 8
.B Quiet \fP( yes | \fBno\fP )
Supress informational messages.  Warning and Error messages will not be
supressed.
.TP 8
.B ReallyQuiet \fP( yes | \fBno\fP )
Supress all messages, including Warning and Error messages.
.TP 8
.B Debug \fP( yes | \fBno\fP )
Print extra debugging information on Warnings and Errors.
.TP 8
.B TimeMe \fP( yes | \fBno\fP )
Force timing information at end of processing.
.TP 8
.B GMTTime \fP( yes | \fBno\fP )
Use \fIGMT \fP(\fIUTC\fP) time instead of local timezone for reports.
.TP 8
.B IgnoreHist \fP( yes | \fBno\fP )
Ignore previous monthly history file.  \fBUSE WITH CAUTION\fP.  Does
not prevent \fIIncremental\fP file processing.
.TP 8
.B FoldSeqErr \fP( yes | \fBno\fP )
Fold out of sequence log records back into analysis by treating them
as if they had the same date/time as the last good record.  Normally,
out of sequence log records are ignored.
.TP 8
.B CountryGraph \fP( \fByes\fP | no )
Display Country Usage Graph in output report.
.TP 8
.B DailyGraph \fP( \fByes\fP | no )
Display Daily Graph in output report.
.TP 8
.B DailyStats \fP( \fByes\fP | no )
Display Daily Statistics in output report.
.TP 8
.B HourlyGraph \fP( \fByes\fP | no )
Display Hourly Graph in output report.
.TP 8
.B HourlyStats \fP( \fByes\fP | no )
Display Hourly Statistics in output report.
.TP 8
.B PageType \fIname\fP
Define the file extensions to consider as a \fIpage\fP.  If a file
is found to have the same extension as \fIname\fP, it will be counted
as a \fIpage\fP (sometimes called a \fIpageview\fP).
.TP 8
.B GraphLegend \fP( \fByes\fP | no )
Allows the color coded graph legends to be enabled/disabled.
.TP 8
.B GraphLines \fInum\fP
Specify the number of background reference lines displayed on the
graphs produced.  Disable by using zero ('\fB0\fP'), default is \fB2\fP.
.TP 8
.B VisitTimeout \fInum\fP
Specifies the visit timeout value.  Default is \fI1800 seconds\fP (30
minutes).  A visit is determined by looking at the difference in time
between the current and last request from a specific site.  If the
difference is greater or equal to the timeout value, the request is
counted as a new visit.  Specified in seconds.
.TP 8
.B IndexAlias \fIname\fP
Use \fIname\fP as an additional alias for \fIindex.*\fP.
.TP 8
.B MangleAgents \fInum\fP
Mangle user agent names based on mangle level \fInum\fP.  See the
\fI-M\fP command line switch for mangle levels and their meaning.
The default is \fB0\fP, which doesn't mangle user agents at all.
.TP 8
.B SearchEngine\fP \fIname\fP \fIvariable\fP
Allows the specification of search engines and their query strings.
The \fIname\fP is the name to match against the referrer string for

CHANGES
--------------------------------------------------------------------
3.10.3.2 [23-Jun-2013] -- Stone Steps Inc. (www.stonesteps.ca)
--------------------------------------------------------------------

 * Added GraphType to the configuration (XML reports)
 * Colors defined in the configuration are propagated now to the XML reports and picked up by XSL
 * Fixed a bug causing the first log record of each file processed twice
 * Made W3C a standalone log file type to account for differences with IIS log file format

--------------------------------------------------------------------
3.10.2.5 [12-Feb-2012] -- Stone Steps Inc. (www.stonesteps.ca)
--------------------------------------------------------------------

 * Reworked how multiple log files are handled to prevent too many open files
 * Added support for IE8 and IE9 in the XML report

--------------------------------------------------------------------
3.10.1.4 [28-Nov-2011] -- Stone Steps Inc. (www.stonesteps.ca)
--------------------------------------------------------------------

 * Added --pipe-log-names to read multiple log file names from the standard input
 * [BUG] Fixed a crash in the XML report generator when TopCountries is set to zero
 * [BUG] Fixed a crash when processing an empty log file
 
--------------------------------------------------------------------
3.10.0.7 [07-Aug-2011] -- Stone Steps Inc. (www.stonesteps.ca)
--------------------------------------------------------------------

 * Added support for multiple log file processing
 * History file is updated when aa report is generated from a database
 * Updated help text to output all options
 * [Windows] Updated Berkeley DB to v5.1.19

--------------------------------------------------------------------
3.9.0.3 [21-Nov-2010] -- Stone Steps Inc. (www.stonesteps.ca)
--------------------------------------------------------------------

 * Implemented time ranges to deal with multiple DST times
 * Added visits to referrers and search string reports
 * Removed IgnoreReferrerRedirect as redundant
 * [BUG] Fixed a bug in the IIS log parser that caused last log field ignored
 * [BUG] Fixed incorrect 3rd party library project references
 * [BUG] Fixed incorrect BDB sequence references between tables
 * [BUG] Fixed a crash bug when a non-numeric VisitTimeout value was specified

--------------------------------------------------------------------
3.8.0.4 [18-Jul-2009] -- Stone Steps Inc. (www.stonesteps.ca)
--------------------------------------------------------------------

    * Added a configuration parameter to control maximum visit length (MaxVisitLength)
    * Added a command line option to print database information (--db-info)
    * Added data type size and byte order indicators to the database system node
    * [Linux] Fixed the source to compile with GCC x64
    * [Linux] Removed -m32 from the compiler and linker options
    * [BUG] Fixed a bug causing visits spanning more than two days lose hits

--------------------------------------------------------------------
3.7.2.3 [26-Apr-2009] -- Stone Steps Inc. (www.stonesteps.ca)
--------------------------------------------------------------------

    * [BUG] Non-Latin characters in URLs, user agents, etc break XML reports
    * [BUG] Time zone names were not encoded in UTF-8 in XML reports
    * [BUG] Fixed incorrect memory allocation when initializing XML help

--------------------------------------------------------------------
3.7.1.3 [18-Apr-2009] -- Stone Steps Inc. (www.stonesteps.ca)
--------------------------------------------------------------------

    * Added handling domain names in the IP address field (AcceptHostNames)
    * Added context-sensitive help to the XML reports (hits, pages, visits, etc)
    * Removed spammer hosts and visits from the conversion totals
    * Added human totals to the monthly summary report
    * Added spammer totals to the monthly summary report
    * [BUG] Changed Max to Total in the Time column of the top downloads report

--------------------------------------------------------------------
3.6.2.1 [19-Feb-2009] -- Stone Steps Inc. (www.stonesteps.ca)
--------------------------------------------------------------------

    * Added a configuration parameter to specify a log directory (LogDir)
    * [BUG] Country report includes spammer visits and the pie chart is generated as if it does not

--------------------------------------------------------------------
3.6.1.7 [11-Jan-2009] -- Stone Steps Inc. (www.stonesteps.ca)
--------------------------------------------------------------------

    * Processing HideX parameters during report generating, so it can be used with --prepare-report
    * Implemented View All link in the top item XML reports
    * Implemented Ctrl-Alt-Home/End/Up/Down navigation between reports
    * Added MaxX (e.g. MaxHosts) to control maximum number of rows in all item reports
    * Added a transfer amount column to the user agent report

--------------------------------------------------------------------
3.5.1.1 [02-Nov-2008] -- Stone Steps Inc. (www.stonesteps.ca)
--------------------------------------------------------------------

    * Added a few missing language variables
    * Improved report generation speed for large databases
    * Added a command line switch to compact the state database (--compact-db)
    * Added a command line switch to finalize the current month (--end-month)
    * Improved spam referrer identification logic
    * Generating reports in XML/XSL, with three graph samples
    * [BUG] Fixed visit count accuracy if there are too many spam referrers
    * [BUG] Dump flags are ignored if there is no explicit output format

--------------------------------------------------------------------
3.4.4.1 [04-Aug-2008] -- Stone Steps Inc. (www.stonesteps.ca)
--------------------------------------------------------------------

    * Added more XML/XSL reports (Top URLs, Top Entry/Exit Pages, Top Downloads)

webalizer.js
/*
   webalizer - a web server log analysis program

   Copyright (c) 2004-1013, Stone Steps Inc. (www.stonesteps.ca)

   See COPYING and Copyright files for additional licensing and copyright information     
   
	webalizer.js
*/
var ELEMENT_NODE  = 1;        // DOM element node type
var TEXT_NODE     = 3;

var KEY_ESC       = 27;
var KEY_HOME      = 36;
var KEY_END       = 35;
var KEY_PAGEUP    = 33;
var KEY_PAGEDOWN  = 34;
var KEY_UP        = 38;
var KEY_DOWN      = 40;
var KEY_LEFT      = 37;
var KEY_RIGHT     = 39;

var PAGE_INDEX    = 0
var PAGE_USAGE    = 1;

var ie6 = false;
var ie7 = false;
var ie8 = false;
var ie9 = false;
var allitems = null;          // instance of ViewAllInfo, if there is an expanded report
var frags = null;             // report identifiers (fragments)
var helpbox = null;           // help box div
var helptext = null;          // hidden help text container div

function ViewAllInfo()
{
   this.rptlink = null;       // a link in front of each report
   this.viewall = null;       // view all link (tbody)
   this.allitems = null;      // all items (tbody)
   this.titlenode = null;     // title table cell (th)
   this.title = null;         // original report title (String)
}

function onloadpage(pagetype)
{
   // check for versions of IE
   ie6 = /MSIE[ \t]+6/.test(window.navigator.userAgent);
   ie7 = /MSIE[ \t]+7/.test(window.navigator.userAgent);
   ie8 = /MSIE[ \t]+8/.test(window.navigator.userAgent);
   ie9 = /MSIE[ \t]+9/.test(window.navigator.userAgent);
   
   // update the fragment array
   if(pagetype == PAGE_USAGE)
      updateFragments();
   
   // find the help box div
   if((helpbox = document.getElementById("helpbox")) != null) 
      helpbox.normalize();
   
   // find the help container div
   helptext = document.getElementById("helptext");
}

function onclickmenu(a)
{
	return true;
}

//
// onpagekeyup
//
// Handles page-level keyboard events
//
function onpagekeyup(event)
{
   // check if it's the ESC key
   if(event.keyCode == KEY_ESC) {
      // if the help is visible, hide it and return
      if(isHelpShown()) {
         hideHelp();
         return;
      }
      
      hideAllItems(true);
   }
   
   // if frags is initialized and
   if(frags) {
      // if Ctrl and Alt are down, process navigation keys
      if(event.ctrlKey && event.altKey) {
         switch(event.keyCode) {
            case KEY_HOME:
               document.location.href = "#" + frags[0];
               break;

            case KEY_END:
               document.location.href = "#" + frags[frags.length-1];
               break;
               
            case KEY_UP:
               document.location.href = "#" + getPrevFragId(getFragId(document.location.href));
               break;

            case KEY_DOWN:
               document.location.href = "#" + getNextFragId(getFragId(document.location.href));
               break;
         }
      }
   }
   
   return true;
}

//
// updateFragments
//
// Populates the fragment array with fragment identifiers from the 
// main menu
//
function updateFragments()
{
   var menu = document.getElementById("main_menu");
   var links;

   // initialize the fragment array
   frags = ["top"];
   
   // index page doesn't have a menu
   if(menu) {
      links = menu.getElementsByTagName("a");
      
      if(links) {
         for(var i = 0; i < links.length; i++)
            frags.push(getFragId(links.item(i).href));
      }
   }
}

//
// getFragId
//
// Returns the fragment identifier in href or an empty string if there 
// is none.
//
function getFragId(href)
{
   var fpos = href? href.lastIndexOf("#") : -1;
   return fpos != -1 ? href.substr(fpos+1) : "";
}

//
// getPrevFragId
//
// Finds fragid in the frags array and returns the previous element. If
// fragid is not found, returns the last element of the array.
//
function getPrevFragId(fragid)
{
   if(fragid) {
      for(var i = 1; i < frags.length; i++) {
         if(frags[i] == fragid)
            return frags[i-1];
      }
   }
   
   return frags[frags.length-1];
}

//
// getNextFragId
//
// Finds fragid in the frags array and returns the next element. If fragid
// is not found, returns the first element of the array.
//
function getNextFragId(fragid)
{
   if(fragid) {
      for(var i = 0; i < frags.length-1; i++) {
         if(frags[i] == fragid)
            return frags[i+1];
      }
   }
   
   return frags[0];
}

//
// hideAllItems
//
// Collapse an expanded all-items section, if there's any
//
function hideAllItems(gotorep)
{
   if(allitems) {
      // make the View All link visible
      allitems.viewall.style.display = (ie6 || ie7 || ie8 || ie9) ? "block" : "table-row-group";
      
      // hide the remaining items
      allitems.allitems.style.display = "none";

      // and restore the report title
      allitems.titlenode.firstChild.data = allitems.title;
      
      // if requested, go to the top of the report
      if(gotorep)
         document.location.href = "#" + allitems.rptlink.name;
         
      // all done, reset the descriptor
      allitems = null;
   }
}

//
// findReportTitle
//
// Given any node inside a report table, returns the report title cell.
// The text data within the cell is normalized before the call returns,
// so it is possible to use firstChild against the returned value to
// access title text.
//
function findReportTitle(node)
{
   // find the parent table
   node = findParentNode(node, "table");
   
   // find the head section
   node = node.firstChild;

   while(node.nodeType != ELEMENT_NODE || node.tagName.toLowerCase() != "thead")
      node = node.nextSibling;

   // find the title row
   node = node.firstChild;

   while(node.nodeType != ELEMENT_NODE || node.tagName.toLowerCase() != "tr" || node.className.indexOf("table_title_tr") == -1)
      node = node.nextSibling;
   
   // find the title cell
   node = node.firstChild;

   while(node.nodeType != ELEMENT_NODE || node.tagName.toLowerCase() != "th")
      node = node.nextSibling;
   
   node.normalize();
   
   return node;
}   

//
// findParentNode
//
// Returns the parent node with the matching tag name of the specified node
//
function findParentNode(node, tagname)
{
   do {
      node = node.parentNode;
   } while(node.nodeType != ELEMENT_NODE || node.tagName.toLowerCase() != tagname);
   
   return node;
}

//
// findNextSibling
//
// Returns the next sibling node with the matching tag name of the specified node
//
function findNextSibling(node, tagname)
{
   do {
      node = node.nextSibling;
   } while(node.nodeType != ELEMENT_NODE || node.tagName.toLowerCase() != tagname);
   
   return node;
}

//
// findPrevSibling
//
// Returns the previous sibling node with the matching tag name of the specified node
//
function findPrevSibling(node, tagname)
{
   do {
      node = node.previousSibling;
   } while(node.nodeType != ELEMENT_NODE || node.tagName.toLowerCase() != tagname);
   
   return node;
}

//
// showAllItems
//
// Shows the hidden table section with the remainder of report items,
// fixes up the title and hides the View All table section.
//
function showAllItems(node, top, count)
{
   if(!node)
      return false;

   // hide the expanded section, if there's any
   hideAllItems(false);
   
   // create an empty array to hold table sections
   allitems = new ViewAllInfo();
   
   //
   // Hide the table section containing the View All link
   //
   node = findParentNode(node, "tbody")

   // store the link node
   allitems.viewall = node;
   
   // and hide the section
   node.style.display = "none";
   
   //
   // Expand the section containing remaining report items
   //
   node = findNextSibling(node, "tbody");

   // store it, so we can use it later to collapse the section
   allitems.allitems = node;

   // and make the section visible (IE6/7 do not understand table-row-group)
   node.style.display = (ie6 || ie7 || ie8 || ie9) ? "block" : "table-row-group";

   //
   // Fix the title to reflect the number of displayed items
   //
   allitems.titlenode = findReportTitle(node);
   
   // keep the title, so we can restore it later
   allitems.title = allitems.titlenode.firstChild.data;
   
   // replace the top number with the new value
   allitems.titlenode.firstChild.data = allitems.title.replace(new RegExp(" " + top + " "), " " + count + " ");

   //
   // We may have collapsed a report that is located before the one we are
   // expanding, which would cause the browser to show some arbitrary part
   // of the markup. In order to avoid this, let's tell the browser to go 
   // to the top of the current report. 
   //
   
   // find the anchor element before the report table
   allitems.rptlink = findPrevSibling(findParentNode(node, "table"), "a");

   // and change the document location
   document.location.href = "#" + allitems.rptlink.name;
   
   return true;
}

//
// findHelpTopic
//
// Walks the help container and returns a copy of the node whose title
// attribute matches topic
//
function findHelpTopic(topic)
{
   var node;

   if(helptext) {
      // start with the first child (may be text node)
      node = helptext.firstChild;
   
      // and loop until we found the topic div
      while(node) {
         if(node.nodeType == ELEMENT_NODE && node.title == topic) {
            // clone the node to keep the source remains intact
            node = node.cloneNode(true);
            
            // remove the title attribute, so the browser doesn't show it as a tool tip
            node.removeAttribute("title");
            
            return node;
         }
            
         node = node.nextSibling;
      }
   }
   
   // return a text node if the topic is not found
   return document.createTextNode("Unknown topic (" + topic + ")");
}

//
// showHelp
//
// Copies text from the title node into the help box title div. Also
// locates help definition div by topic and copies it into the help 
// box.
//
//                                      help topic
//                                      v
// <span onclick="showHelp(event, this, topic)">Hits</span>
// ^                              ^              ^
// title node                     title node     title
//
function showHelp(event, titlenode, topic)
{
   var top, left;
   var title = "";
   var titlediv;
   var scrwidth, scrheight;
   var offsetx, offsety;
   
   // nothing to show
   if(!helpbox || !titlenode)
      return;
   
   // get the text from the title node
   titlenode.normalize();
   title = titlenode.firstChild.data;
   
   // and put it in the title div, after the div with x
   titlediv = helpbox.firstChild;
   
   if(titlediv.firstChild.nextSibling)
      titlediv.replaceChild(document.createTextNode(title), titlediv.firstChild.nextSibling);
   else
      titlediv.appendChild(document.createTextNode(title));
   
   // find a help entry by topic and put it after the title div
   if(titlediv.nextSibling)
      helpbox.replaceChild(findHelpTopic(topic), titlediv.nextSibling);
   else
      helpbox.appendChild(findHelpTopic(topic));

   // place the help box around the mouse pointer
   scrwidth = document.body.parentNode.clientWidth;
   scrheight = document.body.parentNode.clientHeight;
   
   offsetx = (event.clientX > scrwidth/2) ? -350 : 20;
   offsety = (event.clientY > scrheight/2) ? -75 : 20;
   
	top = Math.round(document.body.parentNode.scrollTop + event.clientY + offsety);
	left = Math.round(document.body.parentNode.scrollLeft + (event.clientX + offsetx));
	
	helpbox.style.top = top + "px";
	helpbox.style.left = left + "px";

   // and make it visible
   helpbox.style.display = "block";
}

function hideHelp()
{
   if(helpbox.style.display != "none")
      helpbox.style.display = "none";
}

function isHelpShown()
{
   return helpbox.style.display != "none";
}
